### ISSUE TITLE
No need for factor.com on Windows, can use single factor.exe executable.

### ISSUE THREAD

**Issue by nomennescio** [factor/factor#2200]
On Windows, two executables are build: factor.exe marked as Windows executable, and factor.com marked as Console executable. On the website it is stated:

"_but on Windows, whether or not a program can use the console is a property of the executable file itself, and not of the program's runtime. Two executables are needed; building solely as a console application is unacceptable, because even if a hack is added to the VM's main() function to hide the console as soon as the program starts in graphical mode, it would still flicker on the screen. There is no way around this with the Win32 API._"

That is not true. To explain; under the Windows operating system all executables upon startup are loaded into a separate process, and it is the process that defines the handles for standard output, standard input, and standard error. These handles can be inherited from the parent process. Having a Console is actually independent from having standard file descriptors, and even a GUI application can after startup launch a separate Console (`AllocConsole`, `AttachConsole`) to display output on these descriptors. 

I've build many Windows ("GUI") applications that do not show a GUI, but only show output on standard streams. Using MinGW compilers, you can build an executable with a 'main' or 'wmain' function with the '-mwindows' flag, which will happily run without a Console, but will show standard output in the Terminal of its parent process. The ONLY thing that is NOT working correctly by default, is when running such an executable in 'cmd.exe'. It will run correctly under Cygwin or MSys. An alternative is to call AttachConsole() in code running under cmd.exe, but it is less clean.

I've written such code in the past, but others have since done similar things ([Cygnative](https://github.com/nomennescio/cygnative), [Mintty](https://github.com/mintty/mintty)).
(When not having a Console, one has to be careful on setting up a CTRL-C handler and responding to WM_SHUTDOWN messages)

So to conclude; building factor.com just to have a Console is not needed. For having stdin and stdout in 'cmd.exe' it will work. The same effect can be had in a single executable, which requires code to be added to connect in case running under cmd.exe.

In case factor.com is removed from the build, build times are slightly reduced and distributions will be slightly smaller.

**Comment by mrjbq7** [comment]
Awesome, would love to have some help with this or pointers to example code!

**Comment by nomennescio** [comment]
After some digging, I found this nugget:
https://github.com/Microsoft/Terminal

Microsoft has recently released a true™ Terminal as a 'Console' replacement, and at the same has posted full source code to the Win32 Console Host. I need to dig in to see how it compares to Cygwin/Mingw's mintty. It might be possible to use the same approach for both. In that case Windows users wishing to use stdout/stdin/stderr can be advised to use either terminal, and we could even kill Console support. It would only mean people cannot use standard streams in cmd.exe. I can't imagine a lot of people are using it in the first place.

Alternatively, complex logic could be added to determine if a Console needs to be allocated and do so, but be aware that the Console has real awkwardness when dealing with different encodings which might never be solved satisfactorily. 

(as of today (!) Microsoft has closed its [forum on Command Prompt / Console / Windows Subsystem for Linux](https://wpdev.uservoice.com/forums/266908-command-prompt-console-windows-subsystem-for-l)  and is redirecting everyone to the above site or a github for issues on WSL. That means they've effectively killed Console.)

**Comment by nomennescio** [comment]
As for working code; be aware there's a LOT of misleading and wrong information about using standard streams with the Console. Some articles that point the right direction:

[Death to the Win32 console subsystem](https://www.devever.net/~hl/win32con)
[Stackoverflow](stackoverflow.com/questions/493536/can-one-executable-be-both-a-console-and-gui-application/494000)
[Both GUI and Console](https://www.tillett.info/2013/05/13/how-to-create-a-windows-program-that-works-as-both-as-a-gui-and-console-application/)

but these are still incomplete. When I have time, I will try to dig up my old code that handles some subtleties, but it might be buried on an old HDD. For now I will show how to do this in MinGW:

```
#include <stdio.h>
int main (void) { puts ("Hello World!"); return 0; }
```

compile with the -mwindows flag, run inside Mintty, et voila!

**Comment by nomennescio** [comment]
It still leaves open the question how to properly handle binary/text, codepage/unicode/locale, signals, terminal properties and control etc. in a transparent way.

**Comment by bpollack** [comment]
Woah there, _slow down_. This thread is coming from a good place, but there's a *lot* of confusion.

First, to be clear: `factor.com` is actually a normal EXE file; it's not in COM format. There's a ton of ways to verify this; two seconds in a disassembler or even a hex editor should more than convince you, but I could also point out that COM-format executables are only supported under DOS emulation, so you can just logic out that's not what you're actually looking at.

Second, your notes about console are wrong. Console _the app_ is gradually getting replaced by Terminal _the app_, but this has _nothing_ to do with the Console _subsystem_, which is not going anywhere. In fact, the new Terminal app makes _heavy_ use of massively beefed-up RPC to the console subsystem. I am completely opposed to nuking console support in Factor; that would be a tremendous regression and make me very grumpy.

The only meaningful difference between `factor.com` and `factor.exe` is literally just that the former is tagged console subsystem and the latter is tagged windows subsystem. That's it. You can see that in the `Nmakefile`. Because of this, building both the `.com` and `.exe` is basically free: all the build products have already been made, so you're just spending an extra half a second linking with a different set of executable flags. And you're right that you can open a console from a windows-subsystem, but doing so properly is (as you've correctly noted) significantly more complex than allowing the executable to have two different entry-points that do the right thing and calling it a day, at the expense of having a second, quick linker step and a second executable.

Why then, at the end of the day, is the console app tagged COM and not e.g. `factor-cli-only.exe`? Simple: `.com` files have precedence of `.exe` in the PATH, so if you just type bare `factor` in console, you'll get the console version—which, if you're invoking Factor from inside a console, is probably what you wanted.

So, in summary: I don't agree there's a problem, I _definitely_ don't agree removing console support is a solution, and the proposals so far in this thread seem very complicated compared to the status quo for no benefit I've yet seen.

**Comment by nomennescio** [comment]
Yes, you're right about factor.com not being in the COM format, but I don't think I claimed that here.

As for Console, I was not talking about the subsystem (which indeed will stay), but about the entity launched by AllocConsole and which functions as a terminal, in which e.g. cmd.exe or other executable send their standard input/output. So if you call that the 'app' Console, I'm fine with that.

If you say you're opposed to 'nuking console support in Factor', which console are you talking about?
The only thing I propose is to have standard stream output available in the .exe, so that the .com can go.

From your point of view the benefit is maybe limited, I like to have standard streams available when running in Mintty (Cygwin/Msys), which do NOT work nicely together with -mconsole.

Maybe it's a good idea to first make the requirements on standard streams / Console more clear.

**Comment by AlexIljin** [comment]
>  `factor.com` is actually a normal EXE file; it's not in COM format. There's a ton of ways to verify this;

One simple way is to look at the file size: [COM files](https://en.wikipedia.org/wiki/COM_file) are supposed to be under 64Kb due to the way they were loaded in the early days. The `factor.com` is a normal PE exe-file, only renamed to have the .com extension, and Windows is smart enough to load it correctly anyway.

I run Factor in Windows, and I have the following requirements:
- there should be a GUI executable that starts without the "black window". Typical use case is Listener/Browser and interactive development, where Ctrl-C means "copy to Clipboard";
- there should be a console executable that runs normally from cmd.exe. By "running normally" I mean it uses the existing "black window" for its output; its input and output can be redirected via `<`, `>` and `>>`; `cmd.exe` will be able to pipe (`|`) data to/from it; `cmd.exe` will wait for its termination before continuing on to the next command in cmd/bat-script, and the factor's exit code will be available to `cmd.exe` as `ERRORLEVEL`;
- it should run on Windows 7+.

If you can combine both behaviors in a single executable, I'd be happy to find out about it. I had my own projects, where that would be a nice feature to have, but when I did my research for those projects I found out that it's not possible. Maybe there are some low level tricks that I didn't stumble upon in my research.

PS: I have no intention to install either Cygwin or Msys, and if I release a console tool based on Factor, I can't require my users to install those.

**Comment by bpollack** [comment]
> Yes, you're right about factor.com not being in the COM format, but I don't think I claimed that here.

Correct; you did that [here](https://github.com/factor/factor/issues/2175#issuecomment-526952469), and I wasn't sure if that was a typo or confusion and erred toward the latter.

> As for Console, I was not talking about the subsystem (which indeed will stay), 

Okay, but in that case, your earlier comment above is extremely confusing. You wrote:

> In that case Windows users wishing to use stdout/stdin/stderr can be advised to use either terminal, and we could even kill Console support

Given that Windows Terminal directly uses the Console subsystem and will ultimately be a drop-in replacement for the Windows Console application, I at this point don't know what you were actually trying to say. Could you please clarify?

To be clear: I am not trying to be pedantic here, but I've done Windows-native development since 1995 and if I did _not_ correctly interpret what you're trying to do, then I also openly admit I'm not sure what you actually are attempting to accomplish, and would genuinely appreciate some clarification.

> From your point of view the benefit is maybe limited, I like to have standard streams available when running in Mintty (Cygwin/Msys), which do NOT work nicely together with -mconsole.

Correct: since MinTTY is trying to emulate Xterm, and Cygwin puts in very high effort to pretend to be a *nix, I don't have much interest in having the _Windows version_ of Factor support them well. Both environments have historically had many issues working properly with well-behaved Windows applications, so I'm not particularly interested in bending how Factor works to handle their quirks.

**Comment by nomennescio** [comment]
I've been doing both Unix development and Windows-native development since Windows 3.01, and I've been using Cygwin and Mingw shortly after they've launched. I know that what I'm proposing is very non-standard, and I arrived at a solution after much experimentation. When Mingw switched to Mintty, I stopped using fancy stuff, because it sufficed for what I wanted to accomplish. 

As for the issues with "well-behaved" Windows applications, from experience I can tell that for current Cygwin and Msys2, I cannot think of any issue that I found blocking.

**Comment by nomennescio** [comment]
Maybe one smaller step forward would be to leave factor.com as is: an application compiled with the console flag to be run in the Console subsystem, and adapt factor.exe to use standard streams supplied by the parent processes (see the above code how trivial that is) without launching a Console, and run in the Win32 subsystem. If no standard streams are available, stuff will end up in dev/null anyway.

**Comment by nomennescio** [comment]
With that said, I will close this issue for the time being. First goal is to have a Win32 subsystem Factor.exe that will allow standard input/output streams (e.g. for a GUI application to also show logging when started from a terminal). For now, just leave Factor.com as pure Console subsystem executable as there's apparently a current need for it being a Console subsystem executable.
